packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -96,6 +104,224 @@ if (process.env.NODE_ENV === 'development') {
   OverflowContainerContext.displayName = 'OverflowContainerContext';
 }
 
+/**
+ * Hook that returns the number of items that should be visible based on the size of the container element.
+ *
+ * The returned number should be used to render the element with fewer items.
+ *
+ * @private
+ * @param items Items that this element contains.
+ * @param dimension 'horizontal' (default) or 'vertical'
+ * @returns [callback ref to set on container, stateful count of visible items]
+ *
+ * @example
+ * const items = Array(10).fill().map((_, i) => <span>Item {i}</span>);
+ * const [ref, visibleCount] = useOverflow(items);
+ * ...
+ * return (
+ *   <div ref={ref}>
+ *     {items.slice(0, visibleCount)}
+ *   </div>
+ * );
+ */
+const useOverflow = <T extends HTMLElement>(
+  itemsCount: number,
+  orientation: 'horizontal' | 'vertical' = 'horizontal',
+) => {
+  /** `[number, number]` means that we're still guessing. `null` means that we got the correct `visibleCount`. */
+  type GuessRange = [number, number] | null;
+
+  /** First guess of the number of items that overflows. We refine this guess with subsequent renders. */
+  const STARTING_MAX_ITEMS_COUNT = 32;
+
+  const containerRef = React.useRef<T>(null);
+
+  const initialVisibleCount = React.useMemo(
+    () => Math.min(itemsCount, STARTING_MAX_ITEMS_COUNT),
+    [itemsCount],
+  );
+  const initialVisibleCountGuessRange = React.useMemo(
+    () => [0, initialVisibleCount] satisfies GuessRange,
+    [initialVisibleCount],
+  );
+
+  const [visibleCount, _setVisibleCount] =
+    React.useState<number>(initialVisibleCount);
+  const [visibleCountGuessRange, setVisibleCountGuessRange] =
+    React.useState<GuessRange>(initialVisibleCountGuessRange);
+
+  const isStabilized = visibleCountGuessRange == null;
+
+  /**
+   * Ensures that `visibleCount <= itemsCount`
+   */
+  const setVisibleCount = React.useCallback(
+    (setStateAction: React.SetStateAction<typeof visibleCount>) => {
+      _setVisibleCount((prev) => {
+        const newVisibleCount =
+          typeof setStateAction === 'function'
+            ? setStateAction(prev)
+            : setStateAction;
+
+        return Math.min(newVisibleCount, itemsCount);
+      });
+    },
+    [itemsCount],
+  );
```

Comment:
This kind of wrapper makes it confusing to follow the logic, because we can't ever trust that our `setVisibleCount` call is actually setting the value that we passed. It would be better to remove this wrapper, and call `Math.min` at the place where we are setting the state.